service/cziti/dns-proxy.go

```diff
@@ -341,69 +344,117 @@ outer:
 				} else {
 					if len(respChan) == cap(respChan) {
 						log.Warn("respChan will be blocked. If this warning is continuously displayed please report")
+					} else {
+						log.Tracef("Response received, respChan len: %d, cap: %d ", len(respChan), cap(respChan))
 					}
+
 					respChan <- resp[:n]
 				}
 			}
 		}(proxy)
 	}
 
 	reqs := make(map[uint32]*proxiedReq)
+	var dnsReqMutex = &sync.Mutex{}
 
-	log.Debug("Upstream DNS proxy loop begins")
-	for {
-		select {
-		case pr := <-proxiedRequests:
-			id := (uint32(pr.req.Id) << 16) | uint32(pr.req.Question[0].Qtype)
-			reqs[id] = pr
-			b, _ := pr.req.Pack()
-			for _, proxy := range dnsUpstreams {
-				if _, err := proxy.Write(b); err != nil {
-
-					_ = proxy.Close() //first thing - close the proxy connection
-
-					// when this hits - it never seems to recover. throw a panic which will get recovered
-					// via the defer specified above and try to recreate the connections. this is a heavy
-					// handed way of reconnecting to the DNS server but it should be infrequent
-					log.Panicf("failed to proxy DNS to %s %s %v. %v",
-						dns.Type(pr.req.Question[0].Qtype),
-						pr.req.Question[0].Name,
-						proxy.RemoteAddr(),
-						err)
-				} else {
-					log.Tracef("Proxied request sent to %v from ipv%d listener", proxy.RemoteAddr(), pr.ipVer)
+	dnsReqChannel := make(chan struct{})
+	ticker := time.NewTicker(5 * time.Second)
+	defer func() {
+		close(dnsReqChannel)
+		log.Tracef("Exiting dns request handling routines.")
+	}()
+	go func() {
+	CleanUpIteration:
+		for {
+			select {
+			case <-ticker.C:
+				if len(reqs) == 0 {
+					continue CleanUpIteration
+				}
+				// cleanup requests we didn't get answers for
+				now := time.Now()
+				dnsReqMutex.Lock()
```

Comment:
do you think it'd be useful to add a log statement just before and after this lock so we can see how long this took?

Reply:
Added the trace logs